EIPS/eip-00.md

```diff
@@ -0,0 +1,789 @@
+    } else {
+      senderInput := ABI.encodeWithSelector('validateTransaction', tx, tx.hash);
+      retSender, error := evm.Call(
+          from: AA_ENTRY_POINT,
+          to: sender,
+          input: senderInput,
+          gas: validationGasLimit - retDeployer.gasUsed)
+      assert error == nil
+      assert Date.now() <= retSender.validUntil
+      assert Date.now() >= retSender.validAfter
+      assert retSender.isValid
+    }
+}
+
+```
+
+In order to defend from DoS attack vectors, the block builders performing mempool transaction validation SHOULD consider
+the opcode banning and storage access rules described in EIP-9999.
+
+[Block validation](#execution-layer-block-validation) takes roughly the same amount of work as without AA transactions.
+In any case, validation must execute the entire block in order to verify the state change.
+During this execution, it currently verifies signatures, nonces, and gas payment.
+With Account Abstraction, it will also verify that all the validation frames were successful.
+There is a slight increase in required memory mostly used to store the `context` value that is passed from
+the `paymaster` validation frame to its post-transaction frame.
+
+As long as all transaction validation steps return correct values the block is considered valid.
+Block builders who are willing to relax the rules applied to the validation frames MAY do so.
+
+Such transactions MUST NOT be propagated through the default transaction mempool as they will be rejected by the nodes
+and the sending node will be blocked as a spammer.
+They may be propagated in the alternative mempool that allows them explicitly as defined in EIP-9999.
+
+### All validation state changes apply before all execution ones
+
+Filling a block with AA transactions must not be a challenge for the block builder.
+However, if each transaction during its execution can alter any state that affects the validity of another transaction
+in the mempool, the block builder will be forced to revalidate all transactions in the mempool after each inclusion.
+
+We mitigate that by applying all changes in all the validation frames of a sequence of AA transactions first
+and all execution frames apply immediately after that.
+
+In theory, the validation frames can also invalidate each other, but we define ways to prevent that by applying
+certain rules for the mempool transactions in EIP-9999.
+
+A builder that chooses not to enforce the rules from EIP-9999 **must** take care to re-validate each transaction
+against the mid-block state at the position where it is being included into a block.
+Otherwise, the resulting block is likely to end up being invalid.
+
+### Validation state change virtual transactions
+
+The validation frames of the AA_TX_TYPE transaction are represented as individual virtual transactions by the clients.
+They are assigned their own sequential `transactionIndex`, and their `transactionHash` is defined as
+(`AA_TX_TYPE transaction hash + 1`).
```

Comment:
does this affect the tx root?

Reply:
also why does this need to happen? why can't it just be related to the original transaction hash - e.g. extend eth_getTransactionByHash to return an object `stages` or something which has info about the validation frames

Reply:
> also why does this need to happen? why can't it just be related to the original transaction hash - e.g. extend eth_getTransactionByHash to return an object `stages` or something which has info about the validation frames

The reason we added the artificial 'validation tx hash' is so that clients can observe the "outcomes" for the validation phase separately.

We believe that we do need to provide access to this part of the transaction rather than hiding state changes that happened during validation. One of the L2 AA implementations seems to do this, and it makes debugging hard when a state change during validation has an effect on execution. There is a state change that just magically happens.

We can't concatenate it into one transaction because it is not atomic and there may be state transitions between the two, that would affect the execution transaction.  E.g. `tx1.validation` and `tx2.validation` both use the same `Paymaster`, then `tx1.execution` accesses the `Paymaster` balance, which is changed in `tx2.validation`.
From EVM perspective these are two separate transactions. The consensus just enforces a dependency between them.

At the API level, we could extend `eth_getTransactionByHash` to return two objects as you have suggested, but we can't merge it into one object.

At the block level, it looks like two transactions but at the API level, we can abstract it away.